src/app/qgsapplayertreeviewmenuprovider.cpp

```diff
@@ -217,6 +219,74 @@ QMenu *QgsAppLayerTreeViewMenuProvider::createContextMenu()
         }
       }
 
+      // No raster support in createSqlVectorLayer (yet)
+      if ( vlayer )
+      {
+        std::unique_ptr< QgsAbstractDatabaseProviderConnection> conn { QgsMapLayerUtils::databaseConnection( layer ) };
+        if ( conn )
+          menu->addAction( QgsApplication::getThemeIcon( QStringLiteral( "/dbmanager.svg" ) ), tr( "Update SQL Layer…" ), menu, [ layer ]
+        {
+          std::unique_ptr< QgsAbstractDatabaseProviderConnection> conn2 { QgsMapLayerUtils::databaseConnection( layer ) };
+          if ( conn2 )
+          {
+            QgsDialog dialog;
+            dialog.setObjectName( QStringLiteral( "SqlUpdateDialog" ) );
+            dialog.setWindowTitle( tr( "%1 — Update SQL" ).arg( layer->name() ) );
+            QgsGui::enableAutoGeometryRestore( &dialog );
+            QgsAbstractDatabaseProviderConnection::SqlVectorLayerOptions options { conn2->sqlOptions( layer->source() ) };
+            options.layerName = layer->name();
+            QgsQueryResultWidget *queryResultWidget { new QgsQueryResultWidget( &dialog, conn2.release() ) };
+            queryResultWidget->setUpdateSqlLayerMode( true );
+            queryResultWidget->setSqlVectorLayerOptions( options );
+            queryResultWidget->executeQuery();
+            queryResultWidget->layout()->setMargin( 0 );
+            dialog.layout()->addWidget( queryResultWidget );
+
+            connect( queryResultWidget, &QgsQueryResultWidget::createSqlVectorLayer, queryResultWidget, [queryResultWidget, layer ]( const QString &, const QString &, const QgsAbstractDatabaseProviderConnection::SqlVectorLayerOptions & options )
+            {
+              std::unique_ptr< QgsAbstractDatabaseProviderConnection> conn3 { QgsMapLayerUtils::databaseConnection( layer ) };
+              if ( conn3 )
+              {
+                try
+                {
+                  std::unique_ptr<QgsMapLayer> sqlLayer { conn3->createSqlVectorLayer( options ) };
```

Comment:
This looks like a rather expensive way to update the existing layer -- e.g. it will require that the extent of the query layer is calculated twice (once for the temporary layer construction, and once when you call setDataSource on the actual target layer). Is this avoidable?

Reply:
I'm afraid not: the idea is to not update if the layer is not valid and I cannot think of another way of doing that in advance.

From a UX perspective this is what a user probably wanted: if she tries to update an layer and you make it invalid, she gets an error message inside the widget message bar and the layer is NOT updated.

Do you have any idea?

Reply:
Note that ::createSqlVectorLayer methods already use `QgsVectorLayer::LayerOptions vectorLayerOptions { false, true };` and `vectorLayerOptions.skipCrsValidation = true;`

Reply:
> I'm afraid not: the idea is to not update if the layer is not valid and I cannot think of another way of doing that in advance

Hm, very good point. I can't see a nice way around this either.